app/src/org/commcare/mediadownload/MissingMediaDownloadHelper.kt

```diff
@@ -0,0 +1,196 @@
+package org.commcare.mediadownload
+
+import android.content.Context
+import android.widget.ImageButton
+import android.widget.Toast
+import androidx.work.*
+import kotlinx.coroutines.*
+import org.commcare.CommCareApplication
+import org.commcare.dalvik.R
+import org.commcare.engine.resource.AndroidResourceUtils
+import org.commcare.engine.resource.ResourceInstallUtils
+import org.commcare.preferences.HiddenPreferences
+import org.commcare.resources.model.*
+import org.commcare.utils.AndroidCommCarePlatform
+import org.commcare.utils.AndroidUtil
+import org.commcare.utils.FileUtil
+import org.commcare.views.dialogs.PinnedNotificationWithProgress
+import org.javarosa.core.services.Logger
+import org.javarosa.core.util.SizeBoundUniqueVector
+import java.util.*
+import java.util.concurrent.TimeUnit
+import kotlin.collections.ArrayList
+
+// Contains helper functions to download lazy or missing media resources
+object MissingMediaDownloadHelper : TableStateListener, InstallCancelled {
+
+    var resourceInProgress: Resource? = null
+        private set
+
+    private val jobs = ArrayList<Job>()
```

Comment:
Shouldn't there be a lock of some sort around each one of these jobs, or is that inherent to the job class?

Reply:
each `job` in `jobs` depends on the synchronized  `recoverResource` function in this class which takes care of protecting against spawning multiple download requests for a resource. The `jobs` itself is only meant to provide a way for cancelling all jobs at once.  So it's fine for it to not be thread safe and have duplicate job entries as a result.